File: tools/snapshot.mjs
// Writes an immutable, timestamped metrics snapshot under runs/<run-id>/.
// Immutability is enforced with a read-only file mode after write — the
// run engine never edits a snapshot once written, only reads it back.
import fs from 'node:fs';
import path from 'node:path';
import { redactDeep } from './lib/redact.mjs';

export function writeSnapshot(runDir, metrics, secretMap = {}) {
  fs.mkdirSync(runDir, { recursive: true });
  const snapshotPath = path.join(runDir, 'snapshot.json');
  const redacted = redactDeep(metrics, secretMap);
  fs.writeFileSync(snapshotPath, JSON.stringify(redacted, null, 2));
  fs.chmodSync(snapshotPath, 0o444);
  return snapshotPath;
}

async function selfTest() {
  const os = await import('node:os');
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'snapshot-test-'));
  const secretMap = { alias1: 'super-secret-value' };
  const metrics = { note: 'token super-secret-value here', nested: { again: 'super-secret-value' } };

  const p = writeSnapshot(path.join(tmp, 'runs/run-1'), metrics, secretMap);
  const written = fs.readFileSync(p, 'utf8');
  const stats = fs.statSync(p);

  const checks = [
    ['snapshot file created', fs.existsSync(p)],
    ['snapshot redacted secret before write', !written.includes('super-secret-value')],
    ['snapshot is read-only (immutable)', (stats.mode & 0o200) === 0],
  ];

  fs.chmodSync(p, 0o666); // allow cleanup on Windows
  fs.rmSync(tmp, { recursive: true, force: true });

  let failed = 0;
  for (const [name, ok] of checks) {
    console.log(`${ok ? 'PASS' : 'FAIL'} - ${name}`);
    if (!ok) failed++;
  }
  process.exit(failed ? 1 : 0);
}

const isMain = process.argv[1] && process.argv[1].replace(/\\/g, '/').endsWith('tools/snapshot.mjs');
if (isMain && process.argv.includes('--verify')) selfTest();

File: tools/snapshot.py
# Writes an immutable, timestamped metrics snapshot under runs/<run-id>/.
# Immutability is enforced with a read-only file mode after write - the
# run engine never edits a snapshot once written, only reads it back.
import json
import os
import stat
import sys

try:
    from .lib.redact import redact_deep
except ImportError:
    from lib.redact import redact_deep


def write_snapshot(run_dir, metrics, secret_map=None):
    secret_map = secret_map or {}
    os.makedirs(run_dir, exist_ok=True)
    snapshot_path = os.path.join(run_dir, "snapshot.json")
    redacted = redact_deep(metrics, secret_map)
    with open(snapshot_path, "w", encoding="utf-8", newline="\n") as f:
        json.dump(redacted, f, indent=2)
    os.chmod(snapshot_path, stat.S_IREAD | stat.S_IRGRP | stat.S_IROTH)
    return snapshot_path


def _self_test():
    import shutil
    import tempfile

    tmp = tempfile.mkdtemp(prefix="snapshot-test-")
    secret_map = {"alias1": "super-secret-value"}
    metrics = {"note": "token super-secret-value here", "nested": {"again": "super-secret-value"}}

    p = write_snapshot(os.path.join(tmp, "runs", "run-1"), metrics, secret_map)
    with open(p, "r", encoding="utf-8") as f:
        written = f.read()
    mode = os.stat(p).st_mode

    checks = [
        ("snapshot file created", os.path.exists(p)),
        ("snapshot redacted secret before write", "super-secret-value" not in written),
        ("snapshot is read-only (immutable)", not (mode & stat.S_IWRITE)),
    ]

    os.chmod(p, stat.S_IWRITE | stat.S_IREAD)  # allow cleanup
    shutil.rmtree(tmp, ignore_errors=True)

    failed = 0
    for name, ok in checks:
        print(f"{'PASS' if ok else 'FAIL'} - {name}")
        if not ok:
            failed += 1
    sys.exit(1 if failed else 0)


if __name__ == "__main__" and "--verify" in sys.argv:
    _self_test()
